File: Data Structures & Algorithms/max-area-of-island/submission-0.java
class Solution {
     private static final int[][] directions = {{1, 0}, {-1, 0},
                                               {0, 1}, {0, -1}};
    public int maxAreaOfIsland(int[][] grid) {
        int row= grid.length;
        int col= grid[0].length;
        int area=0;
        for(int i=0; i<row; i++){
            for(int j=0; j<col; j++){
                if(grid[i][j]==1){
                    area= Math.max(area, dfs(i, j, grid));
                }
            }
        }
        return area;
        
    }
    public int dfs(int r, int c, int[][] grid){
        if(r<0||r>=grid.length || c<0 ||c>=grid[0].length|| grid[r][c]==0){
            return 0;
        }
        int res=1;
        grid[r][c]=0;
        for(int[] dir: directions){
            int nr=r+dir[0];
            int nc= c+dir[1];
            res+= dfs(nr, nc, grid);
        }
        return res;
    }
}

File: Data Structures & Algorithms/interleaving-string/submission-0.cpp
class Solution {
public:
vector<vector<int>>dp;
    bool isInterleave(string s1, string s2, string s3) {
        int m= s1.length();
        int n= s2.length();
        if(m+n!= s3.length()){
            return false;
        }
        dp= vector<vector<int>>(m+1, vector<int>(n+1, -1));
        return dfs(0,0,0,s1,s2,s3);
    }
    bool dfs(int i, int j, int k, string s1, string s2, string s3 ){
        if(k==s3.length()){
            return (i==s1.length())&& (j==s2.length());
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        bool res=false;
        if(i<s1.length()&& s1[i]==s3[k]){
            res= dfs(i+1, j, k+1, s1,s2,s3);
        }
        if(!res && j<s2.length()&& s2[j]==s3[k]){
            res=dfs(i, j+1, k+1, s1,s2,s3);
        }
        dp[i][j]=res;
        return res;
    }
};
